aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -327,27 +332,72 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
         }
     );
 
+    if (r.scalar_type() != commonDtype) {
+      r_values = r_values.to(r.scalar_type());
+    }
+    get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
     get_sparse_impl(r)->set_nnz_and_narrow(r_i);
+
     // TODO: I think it may be possible to track inside the loop and
     // detect when we are uncoalesced (e.g., by observing that an
     // index goes backwards) which may be more precise than using the
     // coalesced flag here.  But this is easy.
-    return r._coalesced_(t_coalesced && s_coalesced);
-  } else {
+    return r._coalesced_(coalesced);
+}
+
+SparseTensor& add_out_sparse_non_contiguous(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value, ScalarType commonDtype) {
+    Tensor t_values = t._values().to(commonDtype);
+    Tensor s_values = src._values().to(commonDtype);
+
     // If `t` or `src` contains non-contiguous `values`, `THBlas_axpy` doesn't work
     // and we concat the indices and values tensors instead.
     AT_DISPATCH_ALL_TYPES(
-      s_values.scalar_type(), "add_out_sparse_cuda", [&] {
+      commonDtype, "add_out_sparse_cuda", [&] {
```

Comment:
This is... CPU, right? Is the name here wrong?

Reply:
yup, existing name is wrong, will fix it.